addons/promote/common/src/main/java/org/commonjava/indy/promote/validate/PromotionValidationTools.java

```diff
@@ -636,21 +657,36 @@ public Transfer getTransfer( final StoreResource resource )
 
     private <T> void runParallelAndWait( Collection<T> runCollection, Closure closure, Logger logger )
     {
-        Set<T> todo = new HashSet<>( runCollection );
-        final CountDownLatch latch = new CountDownLatch( todo.size() );
-        todo.forEach( e -> ruleParallelExecutor.execute( () -> {
-            try
-            {
-                logger.trace( "The paralleled exe on element {}", e );
-                closure.call( e );
-            }
-            finally
-            {
-                latch.countDown();
-            }
-        } ) );
+        ThreadContext ctx = ThreadContext.getContext( true );
+        AtomicInteger depth = (AtomicInteger) ctx.computeIfAbsent( ITERATION_DEPTH, k -> new AtomicInteger( -1 ) );
+        depth.incrementAndGet();
```

Comment:
Same as above.

Reply:
I got it wrong up above, but if you look down below you'll see:

```
    MDC.put(ITERATION_DEPTH, String.valueOf( depth.get() ) );
```

The depth AtomicInteger is stored in the ThreadContext, and should be incremented in case we have nested iteration. The idea was to help trace where in rule validation a particular log message was coming from. I decrement it later, after this call, to keep the depth up to date.